File: onion-architecture/java/src/test/java/com/example/infrastructure/adapter/CourseRepositoryImplTest.java
package com.example.infrastructure.adapter;

import java.util.Optional;

import static org.junit.jupiter.api.Assertions.assertEquals;
import static org.junit.jupiter.api.Assertions.assertFalse;
import static org.junit.jupiter.api.Assertions.assertSame;
import static org.junit.jupiter.api.Assertions.assertTrue;
import org.junit.jupiter.api.BeforeEach;
import org.junit.jupiter.api.Test;

import com.example.domain.Course;

class CourseRepositoryImplTest {

    private CourseRepositoryImpl repository;

    @BeforeEach
    void setUp() {
        repository = new CourseRepositoryImpl();
    }

    @Test
    void shouldFindExistingCourseById() {
        Optional<Course> result = repository.findById(1L);

        assertTrue(result.isPresent());
        assertEquals(1L, result.get().getId());
        assertEquals("Java Programming", result.get().getTitle());
    }

    @Test
    void shouldFindAllPreloadedCourses() {
        Optional<Course> javaCourse = repository.findById(1L);
        Optional<Course> springCourse = repository.findById(2L);
        Optional<Course> microservicesCourse = repository.findById(3L);

        assertTrue(javaCourse.isPresent());
        assertEquals("Java Programming", javaCourse.get().getTitle());

        assertTrue(springCourse.isPresent());
        assertEquals("Spring Boot", springCourse.get().getTitle());

        assertTrue(microservicesCourse.isPresent());
        assertEquals("Microservices", microservicesCourse.get().getTitle());
    }

    @Test
    void shouldReturnEmptyOptionalForNonExistentCourse() {
        Optional<Course> result = repository.findById(999L);

        assertFalse(result.isPresent());
    }

    @Test
    void shouldReturnEmptyOptionalForNullId() {
        Optional<Course> result = repository.findById(null);

        assertFalse(result.isPresent());
    }

    @Test
    void shouldReturnSameCourseInstanceForMultipleCalls() {
        Optional<Course> first = repository.findById(1L);
        Optional<Course> second = repository.findById(1L);

        assertTrue(first.isPresent());
        assertTrue(second.isPresent());
        assertSame(first.get(), second.get());
    }

    @Test
    void shouldHavePreloadedCoursesWithCorrectIds() {
        Optional<Course> course1 = repository.findById(1L);
        Optional<Course> course2 = repository.findById(2L);
        Optional<Course> course3 = repository.findById(3L);

        assertTrue(course1.isPresent());
        assertTrue(course2.isPresent());
        assertTrue(course3.isPresent());

        assertEquals(1L, course1.get().getId());
        assertEquals(2L, course2.get().getId());
        assertEquals(3L, course3.get().getId());
    }
}

File: onion-architecture/kotlin/src/test/kotlin/com/example/infrastructure/adapter/CourseRepositoryImplTest.kt
package com.example.infrastructure.adapter

import com.example.domain.Course
import org.junit.jupiter.api.Test
import org.junit.jupiter.api.BeforeEach
import org.junit.jupiter.api.Assertions.*

class CourseRepositoryImplTest {

    private lateinit var repository: CourseRepositoryImpl

    @BeforeEach
    fun setUp() {
        repository = CourseRepositoryImpl()
    }

    @Test
    fun shouldFindExistingCourseById() {
        val result = repository.findById(1L)
        
        assertNotNull(result)
        assertEquals(1L, result!!.id)
        assertEquals("Java Programming", result.title)
    }

    @Test
    fun shouldFindAllPreloadedCourses() {
        val javaCourse = repository.findById(1L)
        val springCourse = repository.findById(2L)
        val microservicesCourse = repository.findById(3L)
        
        assertNotNull(javaCourse)
        assertEquals("Java Programming", javaCourse!!.title)
        
        assertNotNull(springCourse)
        assertEquals("Spring Boot", springCourse!!.title)
        
        assertNotNull(microservicesCourse)
        assertEquals("Microservices", microservicesCourse!!.title)
    }

    @Test
    fun shouldReturnNullForNonExistentCourse() {
        val result = repository.findById(999L)
        
        assertNull(result)
    }

    @Test
    fun shouldReturnNullForZeroId() {
        val result = repository.findById(0L)
        
        assertNull(result)
    }

    @Test
    fun shouldReturnSameCourseInstanceForMultipleCalls() {
        val first = repository.findById(1L)
        val second = repository.findById(1L)
        
        assertNotNull(first)
        assertNotNull(second)
        assertSame(first, second)
    }

    @Test
    fun shouldHavePreloadedCoursesWithCorrectIds() {
        val course1 = repository.findById(1L)
        val course2 = repository.findById(2L)
        val course3 = repository.findById(3L)
        
        assertNotNull(course1)
        assertNotNull(course2)
        assertNotNull(course3)
        
        assertEquals(1L, course1!!.id)
        assertEquals(2L, course2!!.id)
        assertEquals(3L, course3!!.id)
    }
}